src/invert/laplace/impls/petsc3damg/petsc3damg.cxx

```diff
@@ -0,0 +1,583 @@
+  std::vector<OffsetInd3D> interpolatedUpElements = {zero.yp(), zero.xp().yp(), zero.xm().yp(),
+						     zero.yp().zp(), zero.yp().zm()},
+    interpolatedDownElements = {zero.ym(), zero.xp().ym(), zero.xm().ym(), zero.ym().zp(),
+				zero.ym().zm()};
+  std::set<OffsetInd3D> interiorStencil = {zero, zero.xp(), zero.xm(),
+					   zero.zp(), zero.zm(),
+					   zero.xp().zp(), zero.xp().zm(),
+					   zero.xm().zp(), zero.xm().zm()},
+    lowerEdgeStencil = interiorStencil, upperEdgeStencil = interiorStencil;
+
+  for (auto& i : interpolatedDownElements) {
+    for (auto& j : interpPattern) {
+      interiorStencil.insert(i + j);
+      upperEdgeStencil.insert(i + j);
+    }
+    lowerEdgeStencil.insert(i);
+  }
+  for (auto& i : interpolatedUpElements) {
+    for (auto& j : interpPattern) {
+      interiorStencil.insert(i + j);
+      lowerEdgeStencil.insert(i + j);
+    }
+    upperEdgeStencil.insert(i);
+  }
+  std::vector<OffsetInd3D> interiorStencilVector(interiorStencil.begin(), interiorStencil.end()),
+    lowerEdgeStencilVector(lowerEdgeStencil.begin(), lowerEdgeStencil.end()),
+    upperEdgeStencilVector(upperEdgeStencil.begin(), upperEdgeStencil.end());
+
+  // If there is a lower y-boundary then create a part of the stencil
+  // for cells immediately adjacent to it.
+  if (lowerYBound.max() - lowerYBound.min() > 0) {
+    stencil.add([localmesh, &lowerYBound](Ind3D ind) -> bool {
+		  return localmesh->ystart == ind.y() && lowerYBound.intersects(ind.x()); },
+      lowerEdgeStencilVector);
+  }
+
+  // If there is an upper y-boundary then create a part of the stencil
+  // for cells immediately adjacent to it.
+  if (upperYBound.max() - upperYBound.min() > 0) {
+    stencil.add([localmesh, &upperYBound](Ind3D ind) -> bool {
+		  return localmesh->yend == ind.y() && upperYBound.intersects(ind.x()); },
+      upperEdgeStencilVector);
+  }
+
+  // Create a part of the stencil for the interior cells. Although the
+  // test here would also pass for the edge-cells immediately adjacent
+  // to upper/lower y-boundaries, because those tests are run first
+  // the cells will be assigned to those regions.
+  stencil.add([localmesh](Ind3D ind) -> bool {
+		return (localmesh->xstart <= ind.x() &&
+			ind.x() <= localmesh->xend &&
+			localmesh->ystart <= ind.y() &&
+			ind.y() <= localmesh->yend &&
+			localmesh->zstart <= ind.z() &&
+			ind.z() <= localmesh->zend); },
+    interiorStencilVector);
+
+  // Add Y boundaries before X boundaries so corners are assigned to
+  // the former.
+
+  // Note that the tests for whether a cell is in the boundary
+  // actually are not exclusive enough. They really only correspond to
+  // whether the cell is in the guard regions. However, the tests are
+  // much simpler to implement this way and other checks will prevent
+  // guard-cells which are not boundaries from having memory
+  // pre-allocated.
+  
+  // Add lower Y boundary.
+  stencil.add([localmesh](Ind3D ind) -> bool {
```

Comment:
It's a tiny bit more verbose, but in C++14 we can now use lambda capture expressions, so the stencil tests could be like:

```cpp
stencil.add([index = localmesh->ystart - 1](Ind3D ind) { return ind.y() == index; }, ...
```

Reply:
I've changed all of the tests, except the one for the interior cells, as this requires six different values from `localmesh`. I didn't think the extra verbosity was worth it in that case.